File: src/main/java/endytkn/randomEvents/eventObservers/MinecraftEventsObservers.java
package endytkn.randomEvents.eventObservers;

import endytkn.randomEvents.utils.Observer;
import net.minecraftforge.event.TickEvent;
import net.minecraftforge.event.entity.living.LivingDeathEvent;
import net.minecraftforge.event.entity.living.LivingDropsEvent;
import net.minecraftforge.event.entity.player.PlayerInteractEvent;
import net.minecraftforge.eventbus.api.IEventBus;

public class MinecraftEventsObservers {
    public static final Observer<LivingDeathEvent> livingDeathObserver = new Observer<>();
    public static final Observer<TickEvent.PlayerTickEvent> playerTickObserver = new Observer<>();
    public static final Observer<TickEvent> tickObserver = new Observer<>();
    public static final Observer<TickEvent.ServerTickEvent> serverTickObserver = new Observer<>();
    public static final Observer<LivingDropsEvent> livingDropsObserver = new Observer<>();
    public static final Observer<PlayerInteractEvent.EntityInteract> playerInteractEntityObserver = new Observer<>();

    private void onLivingDeath(LivingDeathEvent event) {
        notifyObserver(this.livingDeathObserver, event);
    }

    private void onPlayerTickEvent(TickEvent.PlayerTickEvent event) {
        notifyObserver(playerTickObserver, event);
    }

    private void onLivingDropsEvent(LivingDropsEvent event) {
        notifyObserver(livingDropsObserver, event);
    }

    private void onPlayerInteractEntity(PlayerInteractEvent.EntityInteract event) {
        notifyObserver(playerInteractEntityObserver, event);
    }

    private void onTickEvent(TickEvent event) {
        notifyObserver(tickObserver, event);
    }

    private void onServerTickEvent(TickEvent.ServerTickEvent event) {
        notifyObserver(serverTickObserver, event);
    }

    private <T> void notifyObserver(Observer<T> observer, T event) {
        try {
            observer.notify(event);
        } catch (Exception e) {
            e.printStackTrace(); // Log stack trace for better debugging
            System.err.println("Error while invoking observer for: " + event.getClass().getSimpleName());
        }
    }

    public void register(IEventBus eventBus) {
        eventBus.addListener(this::onServerTickEvent);
        eventBus.addListener(this::onTickEvent);
        eventBus.addListener(this::onPlayerTickEvent);
        eventBus.addListener(this::onLivingDropsEvent);
        eventBus.addListener(this::onPlayerInteractEntity);
        eventBus.addListener(this::onLivingDeath);
    }
}

File: src/main/kotlin/endytkn/randomEvents/minecraftEventsObservers/MinecraftEventsObservers.kt
package endytkn.randomEvents.minecraftEventsObservers

import endytkn.randomEvents.utils.Observer
import net.minecraftforge.event.TickEvent
import net.minecraftforge.event.entity.living.LivingDeathEvent
import net.minecraftforge.event.entity.living.LivingDropsEvent
import net.minecraftforge.event.entity.player.PlayerInteractEvent
import net.minecraftforge.eventbus.api.SubscribeEvent
import net.minecraftforge.fml.common.Mod
import org.jetbrains.annotations.Debug

@Mod.EventBusSubscriber
object MinecraftEventsObservers {
    val onLivingDeathEventObserver = Observer<LivingDeathEvent>()
    val onPlayerTickEventObserver = Observer<TickEvent.PlayerTickEvent>()
    val onTickEventObserver = Observer<TickEvent>()
    val onServerTickEventObserver = Observer<TickEvent.ServerTickEvent>()
    val onLivingDropsEventObserver = Observer<LivingDropsEvent>()
    val onPlayerInteractEntityEventObserver = Observer<PlayerInteractEvent.EntityInteract>()

    @SubscribeEvent
    fun onLivingDeath(event: LivingDeathEvent) {
        try {
            onLivingDeathEventObserver.invoke(event)
        } catch (e: Error) {
            println(e)
            println("error on invoke")
        }
    }

    @SubscribeEvent
    fun onPlayerTickEvent(event: TickEvent.PlayerTickEvent) {
        try {
            onPlayerTickEventObserver.invoke(event)
        } catch (e: Error) {
            println(e)
            println("error on invoke")
        }
    }

    @SubscribeEvent
    fun onLivingDropsEvent(event: LivingDropsEvent) {
        try {
            onLivingDropsEventObserver.invoke(event)
        } catch (e: Error) {
            println(e)
            println("error on invoke")
        }
    }

    @SubscribeEvent
    fun onPlayerInteractEntity(event: PlayerInteractEvent.EntityInteract) {
        try {
            onPlayerInteractEntityEventObserver.invoke(event)
        } catch (e: Error) {
            println(e)
            println("error on invoke")
        }
    }

    @SubscribeEvent
    fun onTickEvent(event: TickEvent) {
        try {
            onTickEventObserver.invoke(event)
        } catch (e: Error) {
            println(e)
            println("error on invoke")
        }
    }

    @SubscribeEvent
    fun onServerTickEvent(event: TickEvent.ServerTickEvent) {
        try {
            onServerTickEventObserver.invoke(event)
        } catch (e: Error) {
            println(e)
            println("error on invoke")
        }
    }
}